### ISSUE TITLE
CentOS build failing after merging #796

### ISSUE THREAD

**Issue by piponazo** [Exiv2/exiv2#800]
After we merged #796 , some tests started to fail in the CentOS build. @D4N sensed that this could happen and he might have some intuition of why this situation has happened. 

Update after investigation:
The problem comes from the fact that CentOS has a very old very of GCC in which regex is not  available:
```
--> Processing Dependency: gcc = 4.8.5-36.el7_6.1 for package: gcc-c++-4.8.5-36.el7_6.1.x86_64
```

See this [link](https://stackoverflow.com/a/12665408/1977218) for more info.

Possible approaches to deal with this situation:
- Upgrade CentOS on the CI ? I am not sure if were are using the latest version or not.
- Remove CentOS from the Gitlab pipelines, since the GCC version it contains it is not fully C++11 compatible.

**Comment by piponazo** [comment]
Description updated after investigation.

**Comment by nehaljwani** [comment]
I think this is the problem: https://stackoverflow.com/a/12665408/1005215

If you upgrade the compiler on CentOS to >=4.9.0 it works. I have a PR at: https://github.com/Exiv2/exiv2/pull/804 which does this.

**Comment by piponazo** [comment]
Exactly, we reached the same conclusion :wink: . I would like to hear @D4N 's opinion on this anyway before merging #804

**Comment by nehaljwani** [comment]
Can this be closed now?

**Comment by D4N** [comment]
Please keep this open, I'd like to take a look at the current state.

**Comment by D4N** [comment]
I've taken a look at the current state of our CentOS build and it now requires a compiler from RedHat's SCL. This could be a problem for the CentOS maintainers of Exiv2, as they might not be allowed to build exiv2 with that compiler due to ABI constraints (I don't have definite answer on that one).

I see the following options:
1. Drop CentOS completely. Imho not a nice thing to do, the CentOS maintainers have been continuously updating exiv2 and are currently shipping 0.27, despite them being an enterprise LTS distribution. We'll effectively throw them under the bus because of a change that's just modernization.
2. Keep the old regex code in just for CentOS. This would be a viable option as the old code is tested and works. But I guess there was a reason for getting rid of it.
3. Use Boost::Regex for GCC 4.8 and 4.9. Boost::Regex is the basis for the C++ >= 11 regex library. The code differences should be minimal (replace `std` with `boost`), but I know what dragging in Boost feels like. Keeping it for compatibility reasons would be IMHO OK though.

 Thoughts/opinions?

**Comment by piponazo** [comment]
My thoughts:
1. They can continue using 0.27 for months/years as we will not release 0.28 soon, and we plan to keep fixing security bugs in 0.27. For me it would be fine to drop CentOS from the master build, until they update their GCC version. At that moment we could add it back. Anyways, we would still have CentOS in the Gitlab pipelines in the branch `0.27-maintenance`.
2. Also a viable option. Although IMHO it would be a pity to take this step. In #796 @Kicer86 simplified the code by using the new C++11 `regex`. We took the decision of moving to C++11 in master (after some discussion about even moving to C++14), and we should be "loyal" to it. If some distribution has a non-fully C++11 compiler it should be their problem, not ours. 
3. I would not be happy about this solution. Even when it is easier to handle dependencies nowadays with **Conan**, I am against bringing such a big dependency just for using a single utility of the library. It would be like depending on Qt just for using `QString`. 

In summary, I would vote for 1, accept 2 if it is the preferred option for many people, and I would not go for 3. 

Another option would be to keep the current solution until we know for sure it can derive into problems.

**Comment by clanmills** [comment]
Is it possible to use Boost::Regex by simply copying a couple of Boost header files into our code-base?  Then we don't require Boost to be installed.

**Comment by D4N** [comment]
We wouldn't need Boost::Regex for the default build, it would be only
required if you are running gcc 4.8 & 4.9 since their implementation of
<regex> is terribly broken. That means for CentOS one would add the
`boost-regex` package to `ci/install_dependencies.sh` and that's
it.

Note, I haven't tested this yet, so it might not work.

**Comment by clanmills** [comment]
We've seen this before.  There was something wrong with Regex on MinGW, so we "downgrade" grep to be sub-string match.  It's documented in README.md in `5.3 MinGW`  under `MinGW and Regex`.
